src/main/java/jenkins/bouncycastle/api/PEMEncodable.java

```diff
@@ -128,23 +143,51 @@ public static PEMEncodable decode(@Nonnull String pem) throws IOException, Unrec
     @Nonnull
     public static PEMEncodable decode(@Nonnull String pem, @Nullable final char[] passphrase)
             throws IOException, UnrecoverableKeyException {
-        PasswordFinder pwf = null;
-        if (passphrase != null) {
-            pwf = new PasswordFinder() {
-                @Override
-                public char[] getPassword() {
-                    return passphrase;
-                }
-            };
-        }
 
-        PEMReader parser = new PEMReader(new StringReader(pem), pwf);
-        try {
-            return new PEMEncodable(parser.readObject());
-        } catch (PasswordException pwE) {
+        try (PEMParser parser = new PEMParser(new StringReader(pem));) {
+
+            Object object = parser.readObject();
+
+            JcaPEMKeyConverter kConv = new JcaPEMKeyConverter().setProvider("BC");
+
+            // handle supported PEM formats.
+            if (object instanceof PEMEncryptedKeyPair) {
+                if (passphrase != null) {
+                    PEMDecryptorProvider dp = new JcePEMDecryptorProviderBuilder().build(passphrase);
+                    PEMEncryptedKeyPair ekp = (PEMEncryptedKeyPair) object;
+                    return new PEMEncodable(kConv.getKeyPair(ekp.decryptKeyPair(dp)));
+                } else {
+                    throw new UnrecoverableKeyException();
+                }
+            } else if (object instanceof PKCS8EncryptedPrivateKeyInfo) {
+                if (passphrase != null) {
+                    InputDecryptorProvider dp = new JceOpenSSLPKCS8DecryptorProviderBuilder().build(passphrase);
+                    PKCS8EncryptedPrivateKeyInfo epk = (PKCS8EncryptedPrivateKeyInfo) object;
+                    return new PEMEncodable(kConv.getPrivateKey(epk.decryptPrivateKeyInfo(dp)));
+                } else {
+                    throw new UnrecoverableKeyException();
+                }
+            } else if (object instanceof PEMKeyPair) {
```

Comment:
Had to do all the nasty conversions to JCA objects conversions here 😞 . My planning was doing it in the toKeyPair, toPublicKey, etc, bu that would mean checked exceptions also on those methods.